opensrp-path/src/main/java/org/smartregister/path/service/intent/ValidateIntentService.java

```diff
@@ -0,0 +1,153 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.smartregister.domain.Response;
+import org.smartregister.path.R;
+import org.smartregister.path.application.VaccinatorApplication;
+import org.smartregister.repository.EventClientRepository;
+import org.smartregister.service.HTTPAgent;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by keyman on 11/10/2017.
+ */
+public class ValidateIntentService extends IntentService {
+
+    private Context context;
+    private HTTPAgent httpAgent;
+    private static final int FETCH_LIMIT = 100;
+    private static final String VALIDATE_SYNC_PATH = "rest/validate/sync";
+
+    public ValidateIntentService() {
+        super("ValidateIntentService");
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        context = getBaseContext();
+        httpAgent = VaccinatorApplication.getInstance().context().getHttpAgent();
+        return super.onStartCommand(intent, flags, startId);
+
+    }
+
+    @Override
+    protected void onHandleIntent(Intent intent) {
+
+        try {
+
+            int fetchLimit = FETCH_LIMIT;
+            EventClientRepository db = VaccinatorApplication.getInstance().eventClientRepository();
+
+            List<String> clientIds = db.getUnValidatedClientBaseEntityIds(fetchLimit);
+            if (!clientIds.isEmpty()) {
+                fetchLimit -= clientIds.size();
+            }
+
+            List<String> eventIds = new ArrayList<>();
+            if (fetchLimit > 0) {
+                eventIds = db.getUnValidatedEventFormSubmissionIds(fetchLimit);
+            }
+
+            JSONObject request = request(clientIds, eventIds);
+            if (request == null) {
+                return;
+            }
+
+            String baseUrl = VaccinatorApplication.getInstance().context().configuration().dristhiBaseURL();
+            if (baseUrl.endsWith(context.getString(R.string.url_separator))) {
+                baseUrl = baseUrl.substring(0, baseUrl.lastIndexOf(context.getString(R.string.url_separator)));
+            }
+
+            String jsonPayload = request.toString();
+            Response<String> response = httpAgent.postWithJsonResponse(
+                    MessageFormat.format("{0}/{1}",
+                            baseUrl,
+                            VALIDATE_SYNC_PATH),
+                    jsonPayload);
+            if (response.isFailure() || StringUtils.isBlank(response.payload())) {
+                Log.e(getClass().getName(), "Validation sync failed.");
+                return;
+            }
+
+            JSONObject results = new JSONObject(response.payload());
+
+            if (results.has(getString(R.string.clients_key))) {
+                JSONArray inValidClients = results.getJSONArray(getString(R.string.clients_key));
+
+                for (int i = 0; i < inValidClients.length(); i++) {
```

Comment:
Lines 87-97 and 102-112 have a lot of repeated code, can you abstract this in java with something like this:

```java
protected void markValidationStatuses(fn, ids, invalidIds) {
    for (int i = 0; i < invalidIds.length(); i++) {
        String invalidId = invalidIds.getString(i);
        if (ids.contains(invalidId)) {
            ids.remove(invalidId);
        }
        db.call(fn, invalidId, false);
    }

    for (String id : ids) {
        db.call(fn, id, true);
    }
}
```

then called with

```java
markValidationStatuses(markClientValidationStatus, clientIds, inValidClients);
markValidationStatuses(markEventValidationStatus, eventIds, inValidEvents);
```